### PR TITLE
IDVA6-346 IDVA-348 Confirm company and add association success page

### PR DESCRIPTION
**Pull request by bmchugh-ch**
Jira links

https://companieshouse.atlassian.net/browse/IDVA6-348
https://companieshouse.atlassian.net/browse/IDVA6-346

- Adds confirm company and success page.
- Adds api enumerations git submodule
- Updates company authentication to use url param

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/routers/controllers/companyAdded.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import { CompanyAddSuccess } from "../handlers/yourCompanies/companyAddSuccess";
+import * as constants from "../../constants";
+import logger from "../../lib/Logger";
+
+export const companyAdded = async (req: Request, res: Response, next: NextFunction) => {
+    try {
+        const viewData = await new CompanyAddSuccess().execute(req, res);
+        res.render(constants.COMPANY_ADDED_SUCCESS_PAGE_TEMPLATE, viewData);
+    } catch (e) {
+        return next(e);
```
Comment (mmodzelewski-ch) [inline src/routers/controllers/companyAdded.ts:11]: Test coverage indicates that this is not tested

File: src/routers/controllers/confirmCompany.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import * as urlUtils from "../../lib/utils/urlUtils";
+
+export const confirmCompanyGet = async (req: Request, res: Response, next: NextFunction) => {
+    try {
+        const session: Session = req.session as Session;
+        const companyProfile = session.data.extra_data.companyProfile;
+        const viewData = await new ConfirmCorrectCompany().execute(req.t, companyProfile);
+        res.render(constants.CONFIRM_COMPANY_PAGE_TEMPLATE, viewData);
+    } catch (e) {
+        return next(e);
```
Comment (mmodzelewski-ch) [inline src/routers/controllers/confirmCompany.ts:16]: Test coverage indicates that this is not tested

File: src/routers/controllers/confirmCompany.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        const company = session.data.extra_data.companyProfile;
+        const userEmailAddress = getLoggedInUserEmail(req.session);
+        const companyNotActive = company.companyStatus.toLocaleLowerCase() !== constants.COMPANY_STATUS_ACTIVE;
+        const isAlreadyAssociated = await isCompanyAssociatedWithUser(company.companyNumber, userEmailAddress);
+        const associationExists = isAlreadyAssociated === constants.COMPNANY_ASSOCIATED_WITH_USER;
+
+        let nextPageUrl = "";
+
+        if (companyNotActive || associationExists) {
+            nextPageUrl = constants.YOUR_COMPANIES_ADD_COMPANY_URL;
```
Comment (mmodzelewski-ch) [inline src/routers/controllers/confirmCompany.ts:32]: Test coverage indicates that this is not tested

File: src/routers/controllers/confirmCompany.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        } else {
+            nextPageUrl = urlUtils.getUrlWithCompanyNumber(constants.CREATE_TRANSACTION_PATH_FULL, company.companyNumber);
+        }
+        return res.redirect(nextPageUrl);
+    } catch (e) {
+        return next(e);
```
Comment (mmodzelewski-ch) [inline src/routers/controllers/confirmCompany.ts:38]: Test coverage indicates that this is not tested

File: src/services/confirmCompanyService.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        type: lookupCompanyType(companyProfile.type),
+        companyStatus: lookupCompanyStatus(companyProfile.companyStatus),
+        dateOfCreation: toReadableFormat(companyProfile.dateOfCreation),
+        registeredOfficeAddress: registeredOfficeAddress
+    };
+};
+
+export const formatTitleCase = (str: string|undefined): string => {
+    if (!str) {
+        return "";
```
Comment (mmodzelewski-ch) [inline src/services/confirmCompanyService.ts:28]: Test coverage indicates that this is not tested

File: src/services/transactionService.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        httpStatusCode: 201
+    });
+};
+
+export const postTransaction = async (session: Session, companyNumber: string, description: string, reference: string): Promise<Transaction> => {
+    const transaction: Transaction = {
+        companyNumber,
+        reference,
+        description,
+        id: "123"
```
Comment (mmodzelewski-ch) [inline src/services/transactionService.ts:15]: I don't know if it matters, but should the id be a hardcoded value like this?

**Review by mmodzelewski-ch — CHANGES_REQUESTED** [review]

File: Makefile [hunk 7]
```diff
@@ -12,7 +12,7 @@ clean:
 	rm -f ./build.log
 
 .PHONY: build
-build:
+build:  update_submodules
```
[the remark this answers is not in the document]
Comment (bmchugh-ch) [reply Makefile:15]: This seems to work okay, its the same in other repos. This line fixed the issue with the install.

File: docker_start.sh [hunk 8]
```diff
@@ -1,8 +1,8 @@
...
 
-cd ..
 npm i
 PORT=3000
 
 export NODE_PORT=${PORT}
-exec node /opt/dist/server.js -- ${PORT}
\ No newline at end of file
+exec node /opt/server.js -- ${PORT}
+# exec node /opt/bin/www.js -- ${PORT}
```
[the remark this answers is not in the document]
Comment (bmchugh-ch) [reply docker_start.sh]: removed

File: src/lib/Logger.ts [hunk 9]
```diff
@@ -3,6 +3,11 @@ import ApplicationLogger from "@companieshouse/structured-logging-node/lib/Appli
 
 const logger: ApplicationLogger = createLogger(process.env.APP_NAME ?? "");
 
+export const createAndLogError = (description: string): Error => {
```
[the remark this answers is not in the document]
Comment (bmchugh-ch) [reply src/lib/Logger.ts:6]: Added unit test for this

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/routers/controllers/companyAdded.ts (hunk 1 above)
Comment (bmchugh-ch) [reply src/routers/controllers/companyAdded.ts]: I've removed the try { ...} catch(e) next(e) in the async functions and installed express-async-errors instead. We should be able to handle or throw errors in these kind of functions without passing them to next() explicitly.

Still means we will have to deal with those errors in error handler middleware though.

File: src/routers/controllers/confirmCompany.ts (hunk 2 above)
Comment (bmchugh-ch) [reply src/routers/controllers/confirmCompany.ts]: As above

File: src/routers/controllers/confirmCompany.ts (hunk 4 above)
Comment (bmchugh-ch) [reply src/routers/controllers/confirmCompany.ts]: As above

File: src/services/confirmCompanyService.ts (hunk 5 above)
Comment (bmchugh-ch) [reply src/services/confirmCompanyService.ts:28]: This is just returning an empty string if no string is passed, we don't need 100% test code coverage. I've tested the formatting of the api responses in the router test but I can't check if its returning an empty string to the page.

File: src/services/transactionService.ts (hunk 6 above)
Comment (bmchugh-ch) [reply src/services/transactionService.ts:15]: We have not setup CHIPS transactions yet. This is just a placeholder mock return value.

File: Makefile (hunk 7 above)
[the remark this answers is not in the document]
Comment (mmodzelewski-ch) [reply Makefile:15]: If it works then OK.

**Review by mmodzelewski-ch — APPROVED** [review]

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by mmodzelewski-ch — DISMISSED** [review]

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by mmodzelewski-ch — APPROVED** [review]

**Pull request merged**